### PR TITLE
Homework unique ptr

### PR DESCRIPTION
**Pull request by Vicx95**

### CODE REVIEW

File: homework/unique_ptr/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,53 @@
+# Download and unpack googletest at configure time
```
Comment (anraf1001) [inline homework/unique_ptr/CMakeLists.txt:1]: You should add `cmake_minimum_required(VERSION <ver. num>)`

File: homework/unique_ptr/UniquePtr.hpp [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+T* UniquePtr<T>::operator->() const {
+    return resource_;
+}
+
+template <typename T>
+T* UniquePtr<T>::get() {
+    if (resource_) {
+        return resource_;
+    }
+    return nullptr;
```
Comment (anraf1001) [inline homework/unique_ptr/UniquePtr.hpp:73]: If this statement is true it means that `resource_` is nullptr. So you can simply return `resource_`.

**Review by anraf1001 — COMMENTED** [review]
Some comments

**Review by jakjan95 — APPROVED** [review]
Code looks good, just correct stuff mentioned by @anraf1001 and it will be perfect :)

**Review by anraf1001 — APPROVED** [review]

**Review by MaciejSalamonski — APPROVED** [review]
i dont have any comments! Nice code~!

File: .github/workflows/module.yml [hunk 3]
```diff
@@ -0,0 +1,20 @@
+name: C++ with CMake CI
+
+on: [push, pull_request]
+
+jobs:
+  UniquePtr:
```
Comment (ziobron) [inline .github/workflows/module.yml:6]: add a valgrind / asan job to prove that there are no memory leaks

File: homework/unique_ptr/UniquePtr.hpp [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+public:
+    UniquePtr(T* ptr);
+
+    UniquePtr(const UniquePtr<T>&) = delete;
+    UniquePtr<T>& operator=(const UniquePtr<T>&) = delete;
+
+    UniquePtr(UniquePtr<T>&& ptr);
+    UniquePtr<T>& operator=(UniquePtr<T>&& ptr);
+
+    T operator*() const;
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:21]:
```suggestion
    T& operator*() const;
```
This operator should return a reference, not a copy

File: homework/unique_ptr/UniquePtr.hpp [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+template <typename T>
+UniquePtr<T>::UniquePtr(UniquePtr<T>&& ptr)
+    : resource_(nullptr) {
+    resource_ = ptr.resource_;
+    ptr.resource_ = nullptr;
+}
+
+template <typename T>
+UniquePtr<T>& UniquePtr<T>::operator=(UniquePtr<T>&& ptr) {
+    if (!resource_) {
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:46]: This is not a valid behavior. If we have a nullptr we return *this.  You should check if it's identity assignment (a == a)

File: homework/unique_ptr/UniquePtr.hpp [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+
+template <typename T>
+T* UniquePtr<T>::get() {
+    return resource_;
+}
+
+template <typename T>
+T* UniquePtr<T>::release() {
+    T* ptr = resource_;
+    delete resource_;
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:76]: we should not delete in release function. You return a dangling pointer from it. ❗️

File: homework/unique_ptr/UniquePtr.hpp [hunk 7]
```diff
@@ -0,0 +1,92 @@
...
+    resource_ = nullptr;
+    return ptr;
+}
+
+template <typename T>
+void UniquePtr<T>::reset(T* ptr) {
+    if (resource_) {
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.hpp:83]: if not needed. deleting a nullptr is safe.

File: homework/unique_ptr/UniquePtr.ut.cpp [hunk 8]
```diff
@@ -0,0 +1,52 @@
+#include <gtest/gtest.h>
+#include "UniquePtr.hpp"
+
+class MoveTest : public testing::Test {
+public:
+    auto moveTest() {
+        UniquePtr<int> ptr{new int{9}};
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.ut.cpp:7]: There are many magic values in this file. Please fix them.

File: homework/unique_ptr/UniquePtr.ut.cpp [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+    UniquePtr<int> ptr{new int{3}};
+    auto ptrNew = ptr.get();
+    ASSERT_EQ(3, *ptrNew);
+}
+
+TEST(UniquePtrTest, AfterReleaseUniquePtrShouldHaveNullPtrInside) {
+    UniquePtr<int> ptr{new int{4}};
+    ptr.release();
+    ASSERT_EQ(nullptr, ptr.get());
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.ut.cpp:27]: there should be a memory leak. You need to assign a result of release() function to a pointer and delete it.

File: homework/unique_ptr/UniquePtr.ut.cpp [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+TEST(UniquePtrTest, IfUniquePtrDereferenceNullPtrExceptionShouldBeThrown) {
+    UniquePtr<int> ptr{nullptr};
+    ASSERT_THROW(*ptr, NullPtrDereferenceException);
+}
+
+TEST(UniquePtrTest, IfResetIsUsedUniquePtrShouldContainNewValue) {
+    UniquePtr<int> ptr{new int{10}};
+    ptr.reset(new int{100});
+    ASSERT_EQ(100, *ptr);
+}
```
Comment (ziobron) [inline homework/unique_ptr/UniquePtr.ut.cpp:46]: what about a reset without an argument?
`ptr.reset();`

**Review by ziobron — CHANGES_REQUESTED** [review]

**Review by WBigaj — COMMENTED** [review]
I have checked your code and have nothing to add that have not been mentioned by other reviewers.

**Review by ziobron — APPROVED** [review]

**Comment by ziobron** [comment]
{ "requirement_ids": [3784, 1119] }
36 XP za implementację unique_ptra (3784)
3 XP za punktualność

**Pull request closed without merging**